File: src/follow_wall.cpp
/*
Basic Follow Wall Algorithm
Author: Roberto Zegers
Date: 2019-March-27
*/

// ROS Libraries
#include <ros/ros.h>
#include <sensor_msgs/LaserScan.h> // Laser Data
#include <geometry_msgs/Twist.h> // Motor Commands

// ROS Subscriber for laser data and messages type (LaserScan)
ros::Subscriber laser_subscriber;
sensor_msgs::LaserScan laser_msg;

// ROS Publisher for motor commands and message type (Twist)
ros::Publisher motor_command_publisher;
geometry_msgs::Twist motor_command;

// global array to keep track of the min. distance value on each zone
float z[5];
// global variable used to set the robots's driving logic
int state;

// The laser_callback function is called each time laser scan data is received
void laser_callback(const sensor_msgs::LaserScan::ConstPtr &scan_msg) {
  // Read and process laser scan values
  laser_msg = *scan_msg;
  // declare an array of float values to keep track of the laser measurements
  std::vector<float> laser_rays;
  // note: array of laser range values is filled from right to left (CCW)
  // rostopic list shows 45*16 = 720 individual laser rays published
  laser_rays = laser_msg.ranges;
  // the total number of laser rays the laser range finder has
  size_t range_size = laser_rays.size();
  ROS_INFO_ONCE("Number of laser rays: [%zu]", range_size); // for debugging
  // variables to store closest (min) distance values on each zone
  z[0] = laser_msg.range_max;
  z[1] = laser_msg.range_max;
  z[2] = laser_msg.range_max;
  z[3] = laser_msg.range_max;
  z[4] = laser_msg.range_max;

  float range_max = laser_msg.range_min;
  // cycle trough all laser range rays
  for (size_t i = 0; i < range_size; i++) {

    // rays < 144, laser rays to the far right side
    if (i < range_size / 5) {
      // get the smallest (closest) laser range value
      if (laser_rays[i] < z[0]) {
        z[0] = laser_rays[i];
      }
    }
    // rays >= 144 and rays < 288, laser rays to the front-right side
    else if (i >= range_size / 5 && i < range_size * 2 / 5) {
      // get the smallest (closest) laser range value
      if (laser_rays[i] < z[1]) {
        z[1] = laser_rays[i];
      }
    }
    // rays >= 288 and rays < 432, laser rays to the front
    else if (i >= range_size * 2 / 5 && i < range_size * 3 / 5) {
      // get the smallest (closest) laser range value
      if (laser_rays[i] < z[2]) {
        z[2] = laser_rays[i];
      }
    }
    // rays >= 432 and rays < 576, laser rays to the front-left side
    else if (i >= range_size * 3 / 5 && i < range_size * 4 / 5) {
      // get the smallest (closest) laser range value
      if (laser_rays[i] < z[3]) {
        z[3] = laser_rays[i];
      }
    }
    // rays > 576 and rays <= 720, laser rays to the far left side
    else if (i >= range_size * 4 / 5 && i <= range_size) {
      // get the smallest (closest) laser range value
      if (laser_rays[i] < z[4]) {
        z[4] = laser_rays[i];
      }
    } else {
      ROS_ERROR("Ray index not found in range size");
    }
  } // end of for
  // ROS_INFO("Closest object to the far right: [%f]: ", z[0]);
  // ROS_INFO("Closest object to the front-right: [%f]: ", z[1]);
  // ROS_INFO("Closest object to the front: [%f]: ", z[2]);
  // ROS_INFO("Closest object to the front-left: [%f]: ", z[3]);
  // ROS_INFO("Closest object to the far left: [%f]: ", z[4]);
}

/*
Detemine velocity commands for each state and fill in a Twist message
*/

void robot_move(geometry_msgs::Twist &motor_command) {
  // ROS_INFO("Wall follower state: [%d]", state);
  switch (state) {
  case 0:
    // Find a wall: turn CW (right) while moving ahead
    motor_command.linear.x = 0.15;
    motor_command.angular.z = -0.15;
    break;

  case 1:
    // Turn left
    motor_command.linear.x = 0.0;
    motor_command.angular.z = 0.2;
    break;

  case 2:
    // Follow the wall: keep moving straight ahead
    motor_command.angular.z = 0.0;
    motor_command.linear.x = 0.15;
    break;

  case 3:
    // Move slow straight ahead
    motor_command.linear.x = 0.10;
    motor_command.angular.z = 0.0;
    break;

  case 4:
    // Reverse turning left
    motor_command.linear.x = -0.5;
    motor_command.angular.z = 0.2; // pos. value equals turning C
    break;
  }
}

/*
Determine the state of the environment surroundings
and the logic used to drive the robot (using 5 zones)
*/
void drive_logic() {
  // fine tune distance (mt) used to consider a region as blocked by an obstacle
  float d = 0.5;

// logic block 1:
  if (z[0] > d && z[1] > d && z[2] > d && z[3] > d && z[4] > d) {
    ROS_INFO("case 1: no obstacles detected");
    state = 0; // find wall: turn CW and move ahead
  } else if (z[0] > d && z[1] > d && z[2] < d && z[3] > d && z[4] > d) {
    ROS_INFO("case 2: obstacle only in front zone");
    state = 1; // turn left
  } else if (z[0] > d && z[1] < d && z[2] > d && z[3] > d && z[4] > d) {
    ROS_INFO("case 3: obstacle only in front-right zone");
    state = 1; // turn left
  } else if (z[0] > d && z[1] > d && z[2] > d && z[3] < d && z[4] > d) {
    ROS_INFO("case 4: obstacle only in front-left zone");
    state = 0; // find wall: turn CW and move ahead
  } else if (z[0] > d && z[1] < d && z[2] < d && z[3] > d && z[4] > d) {
    ROS_INFO("case 5: obstacle in front-right and front zone");
    state = 1; // turn left
  } else if (z[0] > d && z[1] > d && z[2] < d && z[3] < d && z[4] > d) {
    ROS_INFO("case 6: obstacle in front and front-left zone");
    state = 1; // turn left
  } else if (z[0] > d && z[1] < d && z[2] < d && z[3] < d && z[4] > d) {
    ROS_INFO("case 7: obstacle in front-right, front and front-left zone");
    state = 1; // turn left
  } else if (z[0] > d && z[1] < d && z[2] > d && z[3] < d && z[4] > d) {
    ROS_INFO("case 8: obstacle in front-right and front-left zone");
    state = 3; // move slow straight ahead
  }
  // logic block 2:
  else if (z[0] < d && z[1] > d && z[2] > d && z[3] > d && z[4] > d) {
    ROS_INFO("case 9:  obstacle only in right zone");
    state = 2; // follow the wall: keep moving straight ahead
  } else if (z[0] < d && z[1] > d && z[2] < d && z[3] > d && z[4] > d) {
    ROS_INFO("case 10:  obstacle in right and front zone");
    state = 1; // turn left
  } else if (z[0] < d && z[1] < d && z[2] > d && z[3] > d && z[4] > d) {
    ROS_INFO("case 11: obstacle in right and front-right zone");
    state = 1; // turn left
  } else if (z[0] < d && z[1] > d && z[2] > d && z[3] < d && z[4] > d) {
    ROS_INFO("case 12: obstacle in right and front-left zone");
    state = 3; // move slow straight ahead
  } else if (z[0] < d && z[1] < d && z[2] < d && z[3] > d && z[4] > d) {
    ROS_INFO("case 13: obstacle in right, front-right and front zone");
    state = 1; // turn left
  } else if (z[0] < d && z[1] > d && z[2] < d && z[3] < d && z[4] > d) {
    ROS_INFO("case 14: obstacle in right, front and front-left zone");
    state = 1; // turn left
  } else if (z[0] < d && z[1] < d && z[2] < d && z[3] < d && z[4] > d) {
    ROS_INFO("case 15: obst. in right, front-right, front and front-left zone");
    state = 1; // turn left
  } else if (z[0] < d && z[1] < d && z[2] > d && z[3] < d && z[4] > d) {
    ROS_INFO("case 16: obstacle in right, front-right and front-left zone");
    state = 3; // move slow straight ahead
  }
  // logic block 3:
  else if (z[0] > d && z[1] > d && z[2] > d && z[3] > d && z[4] < d) {
    ROS_INFO("case 17: obstacle only in left zone");
    state = 0; // find wall: turn CW and move ahead
  } else if (z[0] > d && z[1] > d && z[2] < d && z[3] > d && z[4] < d) {
    ROS_INFO("case 18: obstacle in front and left zone");
    state = 0; // find wall: turn CW and move ahead
  } else if (z[0] > d && z[1] < d && z[2] > d && z[3] > d && z[4] < d) {
    ROS_INFO("case 19: obstacle in front-right and left zone");
    state = 3; // move slow straight ahead
  } else if (z[0] > d && z[1] > d && z[2] > d && z[3] < d && z[4] < d) {
    ROS_INFO("case 20: obstacle in front-left and left zone");
    state = 0; // find wall: turn CW and move ahead
  } else if (z[0] > d && z[1] < d && z[2] < d && z[3] > d && z[4] < d) {
    ROS_INFO("case 21: obstacle in front-right, front and left zone");
    state = 0; // find wall: turn CW and move ahead
  } else if (z[0] > d && z[1] > d && z[2] < d && z[3] < d && z[4] < d) {
    ROS_INFO("case 22: obstacle in front, front-left and left zone");
    state = 1; // turn left
  } else if (z[0] > d && z[1] < d && z[2] < d && z[3] < d && z[4] < d) {
    ROS_INFO("case 23: obst. in front-right, front, front-left and left zone");
    state = 1; // turn left
  } else if (z[0] > d && z[1] < d && z[2] > d && z[3] < d && z[4] < d) {
    ROS_INFO("case 24: obstacle in front-right, front-left and left zone");
    state = 3; // move slow straight ahead 
  }
  // logic block 4:
  else if (z[0] < d && z[1] > d && z[2] > d && z[3] > d && z[4] < d) {
    ROS_INFO("case 25: obstacle in right and left zone");
    state = 3; // move slow straight ahead
  } else if (z[0] < d && z[1] > d && z[2] < d && z[3] > d && z[4] < d) {
    ROS_INFO("case 26: obstacle in right, front and left zone");
    state = 1; // turn left
  } else if (z[0] < d && z[1] < d && z[2] > d && z[3] > d && z[4] < d) {
    ROS_INFO("case 27: obstacle in right, front-right and left zone");
    state = 1; // turn left
  } else if (z[0] < d && z[1] > d && z[2] > d && z[3] < d && z[4] < d) {
    ROS_INFO("case 28: obstacle in right, front-left and left zone");
    state = 0; // find wall: turn CW and move ahead
  } else if (z[0] < d && z[1] < d && z[2] < d && z[3] > d && z[4] < d) {
    ROS_INFO("case 29: obstacle in right, front-right, front and left zone");
    state = 1; // turn left
  } else if (z[0] < d && z[1] > d && z[2] < d && z[3] < d && z[4] < d) {
    ROS_INFO("case 30: obstacle in right, front, front-left, and left zone");
    state = 0; // find wall: turn CW and move ahead
  } else if (z[0] < d && z[1] < d && z[2] < d && z[3] < d && z[4] < d) {
    ROS_INFO("case 31: obst. in right, front-right, front, front-left and left zone");
    state = 4; // reverse turning left
  } else if (z[0] < d && z[1] < d && z[2] > d && z[3] < d && z[4] < d) {
    ROS_INFO("case 32: obst. in right, front-right, front-left and left zone");
    state = 3; // move slow straight ahead
  } else {
    ROS_INFO("Unknown case");
  }
}

int main(int argc, char **argv) {
  // Initialize a ROS node, change name if required
  ros::init(argc, argv, "follow_wall_commands");

  // Create a ROS NodeHandle object
  ros::NodeHandle n;

  // Inform ROS master that we will be publishing a message of type
  // geometry_msgs::Twist on the robot actuation topic with a publishing queue
  // size of 100
  motor_command_publisher = n.advertise<geometry_msgs::Twist>("/cmd_vel", 100);

  // Subscribe to the /scan topic and call the laser_callback function
  laser_subscriber = n.subscribe("/scan", 1000, laser_callback);
  ROS_INFO_ONCE("Hello World! Follow Wall Customized Version!");
  // Loop, the laser_callback function is called when new laser messages arrives
  while (ros::ok()) {
    ros::spinOnce(); // call all the callbacks waiting to be called
    // take action: determine the state of the environment surroundings
    // and set the logic used to drive the robot
    drive_logic();
    // Fill in motor command message and publish it
    robot_move(motor_command);
    // Publish motor commands to the robot and wait 10ms
    motor_command_publisher.publish(motor_command);
    usleep(10);
  }
  return 0;
}

File: scripts/follow_wall.py
#! /usr/bin/env python

'''
Author: Marco Antonio Arruda
Source: https://bitbucket.org/theconstructcore/two-wheeled-robot-motion-planning
Modified: Roberto Zegers
'''

# import ros stuff
import rospy
from sensor_msgs.msg import LaserScan
from geometry_msgs.msg import Twist
from nav_msgs.msg import Odometry
from tf import transformations
from std_srvs.srv import *

import math

active_ = False
pub_ = None # defining a global publisher
regions_ = {
        'right': 0,
        'fright': 0,
        'front': 0,
        'fleft': 0,
        'left': 0,
}
# state variable and state dictionary
state_ = 0
state_dict_ = {
    0: 'find the wall',
    1: 'turn left',
    2: 'follow the wall',
}

def wall_follower_switch(req):
    global active_
    active_ = req.data
    res = SetBoolResponse()
    res.success = True
    res.message = 'Done!'
    return res

def clbk_laser(msg):
    global regions_
    # read the minimum values of each region
    # a second min is required to filter out 'inf' values, in that case 10 is used 
    regions_ = {
        'right':  min(min(msg.ranges[0:143]), 10),
        'fright': min(min(msg.ranges[144:287]), 10),
        'front':  min(min(msg.ranges[288:431]), 10),
        'fleft':  min(min(msg.ranges[432:575]), 10),
        'left':   min(min(msg.ranges[576:719]), 10),
    }
    
    #print 'Laser Callback Call!' # for debugging
    #rospy.logdebug('Laser Callback Call!')
    take_action()

def change_state(state):
    global state_, state_dict_
    if state is not state_:
        print 'Wall follower - [%s] - %s' % (state, state_dict_[state])
        state_ = state

def take_action():
    ''' obstacle avoidance logic '''
    global regions_
    regions = regions_
    msg = Twist()
    linear_x = 0
    angular_z = 0
    
    state_description = ''
    
    d = 0.5 # fine tune distance used to consider a region blocked as by an obstacle
    
    if regions['front'] > d and regions['fleft'] > d and regions['fright'] > d and regions['left'] > d and regions['right'] > d:
        state_description = 'case 1 - no obstacle sensed'
        print state_description # for debugging
        change_state(0) # find wall: turn CW and move ahead
    elif regions['front'] < d and regions['fleft'] > d and regions['fright'] > d and regions['left'] > d and regions['right'] > d:
        state_description = 'case 2 - obstacle sensed only in front'
        print state_description # for debugging
        change_state(1) # turn left
    elif regions['front'] > d and regions['fleft'] > d and regions['fright'] < d and regions['left'] > d and regions['right'] > d:
        state_description = 'case 3 - obstacle sensed only in front-right'
        print state_description # for debugging
        change_state(1) # turn left
    elif regions['front'] > d and regions['fleft'] < d and regions['fright'] > d and regions['left'] > d and regions['right'] > d:
        state_description = 'case 4 - obstacle sensed only in front-left'
        print state_description # for debugging
        change_state(0) # find wall: turn CW and move ahead
    elif regions['front'] < d and regions['fleft'] > d and regions['fright'] < d and regions['left'] > d and regions['right'] > d:
        state_description = 'case 5 - obstacle sensed in front and front-right'
        print state_description # for debugging
        change_state(1) # turn left
    elif regions['front'] < d and regions['fleft'] < d and regions['fright'] > d and regions['left'] > d and regions['right'] > d:
        state_description = 'case 6 - obstacle sensed in front and front-left'
        print state_description # for debugging
        change_state(1) # turn left
    elif regions['front'] < d and regions['fleft'] < d and regions['fright'] < d and regions['left'] > d and regions['right'] > d:
        state_description = 'case 7 - obstacle sensed in front and front-left and front-right'
        print state_description # for debugging
        change_state(1) # turn left
    elif regions['front'] > d and regions['fleft'] < d and regions['fright'] < d and regions['left'] > d and regions['right'] > d:
        state_description = 'case 8 - obstacle sensed in front-left and front-right'
        print state_description # for debugging
        change_state(3) # move slow straight ahead

# logic block 2:          
    elif regions['front'] > d and regions['fleft'] > d and regions['fright'] > d and regions['left'] > d and regions['right'] < d:
        state_description = 'case 9 - obstacle sensed only in right'
        print state_description # for debugging
        change_state(2) #follow the wall: keep moving straight ahead
    elif regions['front'] < d and regions['fleft'] > d and regions['fright'] > d and regions['left'] > d and regions['right'] < d:
        state_description = 'case 10 - obstacle sensed in front and right'
        print state_description # for debugging
        change_state(1) # turn left
    elif regions['front'] > d and regions['fleft'] > d and regions['fright'] < d and regions['left'] > d and regions['right'] < d:
        state_description = 'case 11 - obstacle sensed in front-right and right'
        print state_description # for debugging
        change_state(1) # turn left
    elif regions['front'] > d and regions['fleft'] < d and regions['fright'] > d and regions['left'] > d and regions['right'] < d:
        state_description = 'case 12 - obstacle sensed in front-left and right'
        print state_description # for debugging
        change_state(3) # move slow straight ahead
    elif regions['front'] < d and regions['fleft'] > d and regions['fright'] < d and regions['left'] > d and regions['right'] < d:
        state_description = 'case 13 - obstacle sensed in front, front-right and right'
        print state_description # for debugging
        change_state(1) # turn left
    elif regions['front'] < d and regions['fleft'] < d and regions['fright'] > d and regions['left'] > d and regions['right'] < d:
        state_description = 'case 14 - obstacle sensed in front, front-left and right'
        print state_description # for debugging
        change_state(1) # turn left
    elif regions['front'] < d and regions['fleft'] < d and regions['fright'] < d and regions['left'] > d and regions['right'] < d:
        state_description = 'case 15 - obstacle sensed in front, front-left, front-right and right'
        print state_description # for debugging
        change_state(1) # turn left
    elif regions['front'] > d and regions['fleft'] < d and regions['fright'] < d and regions['left'] > d and regions['right'] < d:
        state_description = 'case 16 - obstacle sensed in front-left, front-right and right'
        print state_description # for debugging
        change_state(3) # move slow straight ahead

# logic block 3:  
    elif regions['front'] > d and regions['fleft'] > d and regions['fright'] > d and regions['left'] < d and regions['right'] > d:
        state_description = 'case 17 - obstacle sensed only in left'
        print state_description # for debugging
        change_state(0) # find wall: turn CW and move ahead
    elif regions['front'] < d and regions['fleft'] > d and regions['fright'] > d and regions['left'] < d and regions['right'] > d:
        state_description = 'case 18 - obstacle sensed in front and left'
        print state_description # for debugging
        change_state(0) # find wall: turn CW and move ahead
    elif regions['front'] > d and regions['fleft'] > d and regions['fright'] < d and regions['left'] < d and regions['right'] > d:
        state_description = 'case 19 - obstacle sensed in front-right and left'
        print state_description # for debugging
        change_state(3) # move slow straight ahead
    elif regions['front'] > d and regions['fleft'] < d and regions['fright'] > d and regions['left'] < d and regions['right'] > d:
        state_description = 'case 20 - obstacle sensed in front-left and left'
        print state_description # for debugging
        change_state(0) # find wall: turn CW and move ahead
    elif regions['front'] < d and regions['fleft'] > d and regions['fright'] < d and regions['left'] < d and regions['right'] > d:
        state_description = 'case 21 - obstacle sensed in front, front-right and left'
        print state_description # for debugging
        change_state(0) # find wall: turn CW and move ahead
    elif regions['front'] < d and regions['fleft'] < d and regions['fright'] > d and regions['left'] < d and regions['right'] > d:
        state_description = 'case 22 - obstacle sensed in front, front-left and left'
        print state_description # for debugging
        change_state(1) # turn left
    elif regions['front'] < d and regions['fleft'] < d and regions['fright'] < d and regions['left'] < d and regions['right'] > d:
        state_description = 'case 23 - obstacle sensed in front, front-left, front-right and left'
        print state_description # for debugging
        change_state(1) # turn left
    elif regions['front'] > d and regions['fleft'] < d and regions['fright'] < d and regions['left'] < d and regions['right'] > d:
        state_description = 'case 24 - obstacle sensed in front-left, front-right and left'
        print state_description # for debugging
        change_state(3) # move slow straight ahead

# logic block 4:  
    elif regions['front'] > d and regions['fleft'] > d and regions['fright'] > d and regions['left'] < d and regions['right'] < d:
        state_description = 'case 25 - obstacle sensed in left and right'
        print state_description # for debugging
        change_state(3) # move slow straight ahead
    elif regions['front'] < d and regions['fleft'] > d and regions['fright'] > d and regions['left'] < d and regions['right'] < d:
        state_description = 'case 26 - obstacle sensed in front, left and right'
        print state_description # for debugging
        change_state(1) # turn left
    elif regions['front'] > d and regions['fleft'] > d and regions['fright'] < d and regions['left'] < d and regions['right'] < d:
        state_description = 'case 27 - obstacle sensed in front-right, left and right'
        print state_description # for debugging
        change_state(1) # turn left
    elif regions['front'] > d and regions['fleft'] < d and regions['fright'] > d and regions['left'] < d and regions['right'] < d:
        state_description = 'case 28 - obstacle sensed in front-left, left and right'
        print state_description # for debugging
        change_state(0) # find wall: turn CW and move ahead
    elif regions['front'] < d and regions['fleft'] > d and regions['fright'] < d and regions['left'] < d and regions['right'] < d:
        state_description = 'case 29 - obstacle sensed in front, front-right, left and right'
        print state_description # for debugging
        change_state(1) # turn left
    elif regions['front'] < d and regions['fleft'] < d and regions['fright'] > d and regions['left'] < d and regions['right'] < d:
        state_description = 'case 30 - obstacle sensed in front, front-left, left and right'
        print state_description # for debugging
        change_state(0) # find wall: turn CW and move ahead
    elif regions['front'] < d and regions['fleft'] < d and regions['fright'] < d and regions['left'] < d and regions['right'] < d:
        state_description = 'case 31 - obstacle sensed in front, front-left, front-right, left and right'
        print state_description # for debugging
        change_state(4) # reverse turning left
    elif regions['front'] > d and regions['fleft'] < d and regions['fright'] < d and regions['left'] < d and regions['right'] < d:
        state_description = 'case 32 - obstacle sensed in front-left, front-right, left and right'
        print state_description # for debugging
        change_state(3) # move slow straight ahead
        
    else:
        state_description = 'unknown case'
        print 'unknown case' # for debugging
        rospy.loginfo(regions)

def find_wall():
    twistmsg = Twist()
    twistmsg.linear.x = 0.2
    twistmsg.angular.z = -0.2 # negative value equals turning CW
    return twistmsg

def drive_straight_ahead():
    twistmsg = Twist()
    twistmsg.linear.x = 0.15 
    return twistmsg

def turn_left():
    twistmsg = Twist()
    twistmsg.angular.z = 0.2 # positive value equals turning CCW
    return twistmsg

def follow_the_wall():
    global regions_
    
    twistmsg = Twist()
    twistmsg.linear.x = 0.25
    return twistmsg

def reverse_left():
    twistmsg = Twist()
    twistmsg.linear.x = -0.2
    twistmsg.angular.z = 0.2 # positive value equals turning CCW
    return twistmsg

def main():
    global pub_, active_ # to use this global variables inside main()
    
    rospy.init_node('reading_laser')
    
    pub_ = rospy.Publisher('/cmd_vel', Twist, queue_size=1)
    
    sub = rospy.Subscriber('/scan', LaserScan, clbk_laser)
    
    srv = rospy.Service('wall_follower_switch', SetBool, wall_follower_switch)
    
    rate = rospy.Rate(20)
    while not rospy.is_shutdown():
        #if not active_:
        #    rate.sleep()
        #    continue
        
        msg = Twist()
        if state_ == 0:
            msg = find_wall()
            print 'find wall: turn CW and move ahead'
        elif state_ == 1:
            msg = turn_left()
            print 'turn left'
        elif state_ == 2:
            msg = follow_the_wall()
            print 'follow the wall: keep moving straight ahead'
        elif state_ == 3:
            msg = drive_straight_ahead()
            print 'move slow straight ahead'
        elif state_ == 4:
            msg = reverse_left()
            print 'reverse turning left'
            pass
        
        else:
            rospy.logerr('Unknown state!')
        
        pub_.publish(msg)
        rate.sleep()

if __name__ == '__main__':
    main()
